### DISCUSSION TITLE
[Feature] Propertygrid control

### DISCUSSION THREAD

**Discussion by cirrusone** [CommunityToolkit/Labs-Windows#269] [Ideas]
## Describe the problem this feature would solve

Visual Studio has an excellent propertygrid and it would be good to have a similar control inside our own applications to align with WinUI design. Hand building controls are both time consuming and awkward to align on both vertical and horizontal axis when aligning property name(s) and value(s), existing libraries are often missing features or design out of place.

A propertygrid would give well defined structure to a complex layout and simplify handling of lots of properties. 

The main problem this will solve is a user can create a class with properties and bind to a propertygrid in minutes and not have to rearrange controls each time properties are added or amended.

## Describe the solution

Propertygrid should auto generate from class.
Should allow collection of classes.
Should allow nested properties
Should include main WinUi editors
Property editors will allow custom UserControl / custom Editors
Each property should have a description which displays when row selected
Potential to allow multiple columns (more than 2) in the property grid, eg Height Min Max on same row.
Keyboard support to allow tab across or down based on chosen behaviour
Support for class attributes to define propertygrid variables, eg display name different to property name, group name, sort order, property description.
Include file dialog picker
Include ability to sort and search properties

## Describe alternatives you've considered

Main alternatives are as follows
i) manually building individual controls and labels
ii) use open source option
iii) use 3rd party control library

However, most of these options are only WPF or Winforms and will look out of place against WinUI design.

## Additional context & Screenshots

Example of VS propertygrid showing custom editors, grouping and a simple long list of properties

![Properties1](https://user-images.githubusercontent.com/10389314/82217687-b5fded80-9912-11ea-91d9-dd12c9bc0fe7.png)
![Properties2](https://user-images.githubusercontent.com/10389314/82217690-b7c7b100-9912-11ea-89d6-f304ea2aefc8.png)

Please feel free to add or amend anything here.

**Comment by ghost** [comment]
Hello, 'cirrusone! Thanks for submitting a new feature request. I've automatically added a vote 👍 reaction to help get things started. Other community members can vote to help us prioritize this feature in the future!

**Comment by chiefklondog** [comment]
A couple of other items to think about with the property grid:

Support for multi-edit with different types that either share a root type or share identical properties (i.e. you could have 3 classes all with a string 'Name' property and it could be possible to build the property grid to have editors for just those shared items)

Support both inline and popup collection editing. Being able to have your collection shown in-line in the property grid can be very powerful while the popup collection editor can be more performant or easy to implement

A few other controls to think about:
Vector/Point editors 2d, 3d, 4d (common for things like position x/y/z)
Action properties as buttons
Flag enum editor
Multiline text editors

**Comment by amouhsine** [comment]
is there any news regarding this?

**Comment by lextm** [comment]
Instead of waiting forever on such, [created a custom control as prototype](https://github.com/lextudio/UnoPropertyGrid).
